wandb/sdk/system/assets/interfaces.py

```diff
@@ -0,0 +1,165 @@
+
+    def _publish_telemetry(self, telemetry: "TelemetryRecord") -> None:
+        ...  # pragma: no cover
+
+    def publish_files(self, files_dict: "FilesDict") -> None:
+        ...  # pragma: no cover
+
+
+class MetricsMonitor:
+    """
+    Takes care of collecting, sampling, serializing, and publishing a set of metrics.
+    """
+
+    def __init__(
+        self,
+        metrics: List[Metric],
+        interface: Interface,
+        settings: "SettingsStatic",
+        shutdown_event: synchronize.Event,
+    ) -> None:
+        self.metrics = metrics
+        self._interface = interface
+        self._process: Optional[Union[mp.Process, threading.Thread]] = None
+        self._shutdown_event: synchronize.Event = shutdown_event
+
+        self.sampling_interval: float = float(
+            max(
+                0.1,
+                settings._stats_sample_rate_seconds,
+            )
+        )  # seconds
+        # The number of samples to aggregate (e.g. average or compute max/min etc.)
+        # before publishing; defaults to 15; valid range: [1:30]
+        self.samples_to_aggregate: int = min(
+            30, max(1, settings._stats_samples_to_average)
+        )
+
+    def monitor(self) -> None:
+        """Poll the Asset metrics"""
+        while not self._shutdown_event.is_set():
+            for _ in range(self.samples_to_aggregate):
+                for metric in self.metrics:
+                    try:
+                        metric.sample()
+                    except Exception as e:
+                        wandb.termerror(f"Failed to sample metric: {e}", repeat=False)
+                self._shutdown_event.wait(self.sampling_interval)
+                if self._shutdown_event.is_set():
+                    break
+            self.publish()
+
+    def aggregate(self) -> dict:
+        """Return a dict of metrics"""
+        aggregated_metrics = {}
+        for metric in self.metrics:
+            try:
+                serialized_metric = metric.aggregate()
+                aggregated_metrics.update(serialized_metric)
+                # aggregated_metrics = wandb.util.merge_dicts(
+                #     aggregated_metrics, metric.serialize()
+                # )
+            except Exception as e:
+                wandb.termerror(f"Failed to serialize metric: {e}", repeat=False)
+        return aggregated_metrics
+
+    def publish(self) -> None:
+        """Publish the Asset metrics"""
+        try:
+            aggregated_metrics = self.aggregate()
+            if aggregated_metrics:
+                self._interface.publish_stats(aggregated_metrics)
+            for metric in self.metrics:
+                metric.clear()
+        except Exception as e:
+            wandb.termerror(f"Failed to publish metrics: {e}", repeat=False)
+
+    def start(self) -> None:
+        if self._process is None and not self._shutdown_event.is_set():
+            self._process = threading.Thread(target=self.monitor, daemon=True)
```

Comment:
are these threads being named anywhere?  this would make it hard to debug with unnamed threads

Reply:
done!